src/drivers/lego/driver.rs

```diff
@@ -0,0 +1,628 @@
+        event::{ImuAxisInput, TouchAxisInput, TouchButtonEvent},
+        hid_report::GamepadMode,
+        CLICK_DELAY, PAD_FORCE_NORMAL, RELEASE_DELAY, XINPUT_COMMAND_ID,
+    },
+    input::capability::Capability,
+};
+
+use super::{
+    event::{
+        AxisEvent, BinaryInput, Event, GamepadButtonEvent, JoyAxisInput, MouseWheelInput,
+        TriggerEvent, TriggerInput,
+    },
+    hid_report::XInputDataReport,
+    GP_IID, HID_TIMEOUT, PIDS, VID, XINPUT_DATA, XINPUT_PACKET_SIZE,
+};
+
+pub struct Driver {
+    /// HIDRAW device instance
+    hid_device: HidDevice,
+    /// Udev device instance
+    udev_device: UdevDevice,
+    /// List of events that should not be generated
+    filtered_events: HashSet<Capability>,
+    /// Timestamp of the first touch event.
+    first_touch: Instant,
+    /// Whether or not we are currently holding a click-to-click.
+    is_clicked: bool,
+    /// Whether or not we are detecting a touch event currently.
+    is_touching: bool,
+    /// Timestamp of the last touch event.
+    last_touch: Instant,
+    /// Whether or not a touch event was started that hasn't been cleared.
+    touch_started: bool,
+    /// State for the internal gamepad controller
+    xinput_state: Option<XInputDataReport>,
+}
+
+impl Driver {
+    pub fn new(udev_device: UdevDevice) -> Result<Self, Box<dyn Error + Send + Sync>> {
+        let fmtpath = udev_device.devnode().clone();
+        let path = CString::new(fmtpath.clone())?;
+        let api = hidapi::HidApi::new()?;
+        let hid_device = api.open_path(&path)?;
+        let info = hid_device.get_device_info()?;
+
+        if info.vendor_id() != VID
+            || !PIDS.contains(&info.product_id())
+            || info.interface_number() != GP_IID
+        {
+            return Err(format!("Device '{fmtpath}' is not a Legion Go S Controller").into());
+        }
+
+        Ok(Self {
+            udev_device,
+            hid_device,
+            filtered_events: Default::default(),
+            first_touch: Instant::now(),
+            is_clicked: false,
+            is_touching: false,
+            last_touch: Instant::now(),
+            touch_started: false,
+            xinput_state: None,
+        })
+    }
+
+    pub fn update_filtered_events(&mut self, events: HashSet<Capability>) {
+        self.filtered_events = events;
+    }
+
+    pub fn get_default_event_filter(
+        &self,
+    ) -> Result<HashSet<Capability>, Box<dyn Error + Send + Sync>> {
+        let device = self.udev_device.get_device()?;
+        let Some(parent) = device.parent() else {
+            return Ok(HashSet::from([
+                Capability::Accelerometer(Source::Left),
+                Capability::Accelerometer(Source::Right),
+                Capability::Accelerometer(Source::Center),
+                Capability::Gyroscope(Source::Left),
+                Capability::Gyroscope(Source::Right),
+                Capability::Gyroscope(Source::Center),
+            ]));
+        };
```

Comment:
If you want, you can reduce the amount of duplication here by defining a constant:

```suggestion
        const DEFAULT_FILTER: [Capability; 6] = [
            Capability::Accelerometer(Source::Left),
            Capability::Accelerometer(Source::Right),
            Capability::Accelerometer(Source::Center),
            Capability::Gyroscope(Source::Left),
            Capability::Gyroscope(Source::Right),
            Capability::Gyroscope(Source::Center),
        ];
        let device = self.udev_device.get_device()?;
        let Some(parent) = device.parent() else {
            return Ok(HashSet::from(DEFAULT_FILTER));
        };
```

Reply:
Thanks, I meant to go back and do that